### PR TITLE
manager: Only invoke a single sigchld per unit within a cleanup cycle

### PR DESCRIPTION
**Pull request by kyle-walker**
By default, each iteration of manager_dispatch_sigchld() results in a unit level
sigchld event being invoked. For scope units, this results in a scope_sigchld_event()
which can seemingly stall for workloads that have a large number of PIDs within the
scope. The stall exhibits itself as a SIG_0 being initiated for each u->pids entry
as a result of pid_is_unwaited().

v2:
This patch resolves this condition by only paying to cost of a sigchld in the underlying
scope unit once per sigchld iteration. A new "sigchldgen" member resides within the
Manager and Unit structs. The Manager is incremented once per sigchld iteration, and
the Unit member is set to the same value as the manager each time that a sigchld event
is invoked. If the Manager and Unit members match, the sigchld event is not invoked for
that iteration.

### CODE REVIEW

**Comment by poettering** [comment]
ah, did you see my newer comment on https://github.com/systemd/systemd/pull/3612#issuecomment-229541072? would be cool to reuse the sd-event iteration counter, as suggested there.

**Comment by kyle-walker** [comment]
@poettering 
I'm sorry, I didn't see the update until just a moment ago! I will retool this PR and attempt to reuse that counter instead.

Sorry for the mixup!

**Comment by kyle-walker** [comment]
@poettering 
Here is the latest iteration of this patch. This time around I hooked into your #3631 PR with sd_event_get_iteration() instead of maintaining my own iteration counter. Let me know if there is any difficulties with this revision.

**Comment by poettering** [comment]
the CI failure appears unrelated. Merging.

**Comment by evverx** [comment]
```
src/core/manager.c: In function ‘invoke_sigchld_event’:
src/core/manager.c:1736:35: warning: format ‘%llu’ expects argument of type ‘long long unsigned int’, but argument 8 has type ‘uint64_t {aka long unsigned int}’ [-Wformat=]
                         log_debug("%s already issued a sigchld this iteration %llu, skipping. Pids still being watched %d", u->id, iteration, set_size(u->pids));
                                   ^
./src/basic/log.h:171:82: note: in definition of macro ‘log_full_errno’
                         ? log_internal(_level, _e, __FILE__, __LINE__, __func__, __VA_ARGS__) \
                                                                                  ^~~~~~~~~~~
./src/basic/log.h:178:28: note: in expansion of macro ‘log_full’
 #define log_debug(...)     log_full(LOG_DEBUG,   __VA_ARGS__)
                            ^~~~~~~~
src/core/manager.c:1736:25: note: in expansion of macro ‘log_debug’
                         log_debug("%s already issued a sigchld this iteration %llu, skipping. Pids still being watched %d", u->id, iteration, set_size(u->pids));
                         ^~~~~~~~~
```

I think should be `%" PRIu64 "` instead of `%llu`.

**Comment by evverx** [comment]
> this results in a scope_sigchld_event()
> which can seemingly stall for workloads that have a large number of PIDs within the
> scope. The stall exhibits itself as a SIG_0 being initiated for each u->pids entry
> as a result of pid_is_unwaited().

@disneyworldguy, interesting. Did you measure this stall somehow?

**Comment by kyle-walker** [comment]
@evverx 

> I think should be %" PRIu64 " instead of %llu.

Ahh, missed that. Since this has already been merged, would the process here be a follow-up PR with the printf correction?

> @disneyworldguy, interesting. Did you measure this stall somehow?

Yep, once it became apparent that the number of orphan zombies being created outweighed the number of processes that systemd was reaping, "perf record -g -p 1" and a subsequent "perf trace -p 1".

More targeted probepoints within the tail end of the manager_dispatch_sigchld() loop and another within pid_is_unwaited(), returning the pid passed, showed specifically what was occurring.

**Comment by evverx** [comment]
@disneyworldguy 

> Since this has already been merged, would the process here be a follow-up PR with the printf correction?

Yes. Ping me, I'll merge that PR.

>  "perf record -g -p 1" and a subsequent "perf trace -p 1"

Oh, sorry. I mean, do you have some numbers?
I.e. something like this
- Before:
  2 Zombies/sec
- After:
  100500 Zombies/sec + coffee

:)

**Comment by kyle-walker** [comment]
@evverx 

> Oh, sorry. I mean, do you have some numbers?

Understood. It looks like the usual zombie consumption rate is somewhere around 12k/sec average on my testbed. 

```
#           time             counts unit events
376.303502273             12,456      probe_systemd:manager_dispatch_sigchld                                   
377.315128046             13,170      probe_systemd:manager_dispatch_sigchld                                   
378.317544409             13,373      probe_systemd:manager_dispatch_sigchld                                   
379.318120348             14,075      probe_systemd:manager_dispatch_sigchld                                   
380.320784172             12,544      probe_systemd:manager_dispatch_sigchld                                   
381.322461610             12,142      probe_systemd:manager_dispatch_sigchld                                   
382.324428941             11,202      probe_systemd:manager_dispatch_sigchld                                   
383.328433322             12,595      probe_systemd:manager_dispatch_sigchld                                   
```

The problem here is, unfortunately, a little more insidious than an outright Zombies/sec issue. Within the scope_sigchld_event() function call, it calls unit_tidy_watch_pids(). This happens for each zombie PID returned by the waitid() at the manager_dispatch_sigchld() outset.

What happened previously was that, for small scopes, the amount of time spent in the invoke_sigchld_event() callpath was quite small. However, as the list of processes in the scope grew, it meant that much more time spent in unit_tidy_watch_pids(), for each zombie inbound. In short, at a point where the rate of zombie creation is greater than the time spent invoking each sigchld event in a given scope, systemd gets to a point where it can no longer recover effectively. Worse yet, once it is in that state, there isn't any way to recover outside of a system reboot. Mainly in that if it takes 5 minutes to reap a single zombie due to the scope task list being 500,000 entries large, it will take an extreme interval of time, with no further zombies being created, to clear out the backlog.

This should help the condition greatly, working to evaluate how much now, since we only pay the cost once per iteration at most. Which should keep systemd up towards it's best consumption rate marks when this worst case scenario hits.

I went ahead and filed PR #3640 to fix the printf statement. Thank you for the feedback!

tl;dr When the zombie horde gets out of hand, systemd needs to stop comparing each to all the other ones and just focus on channeling Rick Grimes.
